galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -87,21 +91,72 @@ class CollectionViewSet(api_base.LocalSettingsMixin,
     serializer_class = CollectionSerializer
 
 
-class UnpaginatedCollectionVersionViewSet(api_base.LocalSettingsMixin,
-                                          ViewNamespaceSerializerContextMixin,
-                                          pulp_ansible_views.UnpaginatedCollectionVersionViewSet):
+class UnpaginatedCollectionVersionViewSet(
+    api_base.LocalSettingsMixin,
+    ViewNamespaceSerializerContextMixin,
+    pulp_ansible_views.UnpaginatedCollectionVersionViewSet,
+):
     pagination_class = None
     serializer_class = UnpaginatedCollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
 
 
+def get_dependents(parent):
+    """Given a parent collection version, return a list of
+       collection versions that depend on it.
+    """
+    key = f"{parent.namespace}.{parent.name}"
+    dependents = []
+    for child in CollectionVersion.objects.filter(dependencies__has_key=key):
+        spec = SimpleSpec(child.dependencies[key])
+        if spec.match(Version(parent.version)):
+            dependents.append(child)
+    return dependents
+
+
 class CollectionVersionViewSet(api_base.LocalSettingsMixin,
                                ViewNamespaceSerializerContextMixin,
                                pulp_ansible_views.CollectionVersionViewSet):
     serializer_class = CollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
     list_serializer_class = CollectionVersionListSerializer
 
+    def destroy(self, request: Request, *args, **kwargs) -> Response:
+        """
+        Allow a CollectionVersion to be deleted.
+        1. Perform Dependency Check to verify that the collection version can be deleted
+        2. If the collection version can’t be deleted, return the reason why
+        3. If it can, dispatch task to delete CollectionVersion and clean up repository.
+           If the version being deleted is the last collection version in the collection,
+           remove the collection object as well.
+        """
+        collection_version = self.get_object()
+
+        # dependency check
+        dependents = get_dependents(collection_version)
+        if dependents:
+            return Response(
+                {
+                    "detail": _(
+                        "Collection {namespace}.{name} could not be deleted "
+                        "because there are other collections that require it."
+                    ).format(
+                        namespace=collection_version.namespace,
+                        name=collection_version.collection.name,
+                    ),
+                    "dependent_collection_versions": [f"{dep.namespace}.{dep.name} {dep.version}" for dep in dependents],
```

Comment:
changed from `dependent_collections` to `dependent_collection_versions` since we are returning collection_versions in `get_dependents()` and this helps users track down which versions depend on the collection_version they are trying to delete